### ISSUE TITLE
implement generic ShadowMap in Scala 

### ISSUE THREAD

**Issue by bjornregnell** [avatar1337/ShadowHashMap#2]
Some ideas for implementation:
```
scala> :paste

def reverseMap[K,V](xs:Map[K,V]): Map[V, Vector[K]] = 
  xs.toVector
    .map { case (k,v) => (v,k) }
    .groupBy(_._1)
    .mapValues(_.map(_._2))

scala> val xs = Map(1 -> "gurka", 2 -> "tomat", 3 -> "gurka")
xs: scala.collection.immutable.Map[Int,String] = Map(1 -> gurka, 2 -> tomat, 3 -> gurka)

scala> reverseMap(xs)
res13: Map[String,Vector[Int]] = Map(tomat -> Vector(2), gurka -> Vector(1, 3))

scala> 
```

**Comment by avatar1337** [comment]
Thanks I will try this one out. This however creates two maps, which is what I was trying to avoid. My objective was to create a fast HashMap in both directions without taking up twice as much memory. You could use reverseMap(xs) every time without saving it, as you did, but then the performance would be bad, right? So you have the choice of either using up a lot of memory or bad performance. I could be wrong though, this could have great performance. I will do some testing. =)

**Comment by bjornregnell** [comment]
But you can create the "shadow map" with reversed mappings lazilly...
```
scala> :pa

case class ShadowMap[K, V](underlying: Map[K, V]) {
  private val reversed = scala.collection.mutable.Map.empty[V, Vector[K]]
  def apply(key: K): V = underlying(key)
  def keysOf(value: V): Vector[K] =
    if (reversed.isDefinedAt(value)) reversed(value)
    else {
      val keys = underlying.collect{ case (k, v) if v == value => k}.toVector
      reversed += (value -> keys)
      keys
    }
}
object ShadowMap {
  def apply[K, V](kvs: (K,V)*): ShadowMap[K, V] = new ShadowMap(kvs.toMap)
}

scala> val xs = ShadowMap(1 -> "gurka", 2 -> "tomat", 3 -> "gurka")
xs: ShadowMap[Int,String] = ShadowMap(Map(1 -> gurka, 2 -> tomat, 3 -> gurka))

scala> xs.keysOf("gurka")
res0: Vector[Int] = Vector(1, 3)

```

**Comment by bjornregnell** [comment]
Edit: You should probably make the bag a Map with a value type Set instead of a Vector:
```
case class ShadowMap[K, V](underlying: Map[K, V]) {
  private val reversed = scala.collection.mutable.Map.empty[V, Set[K]]
  def apply(key: K): V = underlying(key)
  def keysOf(value: V): Set[K] =
    if (reversed.isDefinedAt(value)) reversed(value)
    else {
      val keys = underlying.collect{ case (k, v) if v == value => k}.toSet
      reversed += (value -> keys)
      keys
    }
}
object ShadowMap {
  def apply[K, V](kvs: (K,V)*): ShadowMap[K, V] = new ShadowMap(kvs.toMap)
}
```

**Comment by avatar1337** [comment]
Yes but isn't 

> `underlying.collect{ case (k, v) if v == value => k}.toSet`

at best an O(n) operation? If you have a million keys, it would have to go through a million keys in order to check if the value fits. Mine only has to go through the entries that happens to lie at that index. It may have to go through 10 entries instead of a million if only say 2 keys have that value. That's the power of a ShadowHashMap. It has a direct reference to the entries containing that value.
It calculates exactly where the keys should be in the table based on value.

**Comment by bjornregnell** [comment]
Well, if your use case is immutability after build, its not that bad to build it once... But ok, if you want efficient mutability before all elements are there, then probably this is better:
```
class ReversableMapBuilder[K, V] {
  private val underlying = scala.collection.mutable.Map.empty[K, V]   
  private val shadow =  scala.collection.mutable.Map.empty[V,Set[K]]

  def apply(key: K): V = underlying(key)
  def keysOf(value: V): Set[K] = shadow(value)

  def add(kv: (K, V)): Unit = {
    underlying += kv
    shadow += (kv._2 -> (shadow.getOrElse(kv._2, Set.empty[K]) ++ Set(kv._1)))
  }

  def toMap: Map[K, V] = underlying.toMap  //get an immutable map with all its goodies
  def reversed:  Map[V, Set[K]] = shadow.toMap // get an immutable bag 
}

object ReversableMapBuilder {
  def empty[K,V]: ReversableMapBuilder[K, V] = new ReversableMapBuilder
  def apply[K, V](kvs: (K,V)*): ReversableMapBuilder[K, V] = {
    val result = empty[K,V]
    kvs.foreach(result.add)
    result
  }
}


scala> val rmb = ReversableMapBuilder(1 -> "gurka", 2 -> "tomat", 3 -> "gurka")
rmb: ReversableMapBuilder[Int,String] = ReversableMapBuilder@27f61670

scala> rmb.add(5 -> "tomat")

scala> rmb.keysOf("tomat")
res11: Set[Int] = Set(2, 5)

```

**Comment by avatar1337** [comment]
Exactly, however this was what I wanted to avoid hehe. This creates two maps and thus takes up twice as much memory, right? This may be all you need in practice but mine should take up much less memory without compromising speed. Let me show you what my Entry class looks like:

```
private static class Entry<K,V> implements Map.Entry<K, V> {
        private K key;
        private V value;
        private Entry<K,V> shadow;
        private Entry<K,V> next;
        
	public Entry(K key, V value){
                this.key = key;
                this.value = value;
                this.next = null;
                this.shadow = null;
        }
.
.
.
}
```
As you can see my shadow is simply a reference to another Entry. When a pair is added an Entry object is created and added to the table but my add() method also calculates which Entry should contain a shadow reference back to that Entry. Another Entry's shadow attribute will thus contain a reference back to the Entry that was just added. If there is no Entry at that position, an empty one will be created. The next time a real Entry must be created at that position, only the key and value attributes will be updated. When I want to retrieve the keys from the value my getKeys(value) method again calculates where the Entry is containing a shadow attribute that contains a reference back to the Entry containing that value. Bottom line, I use the same trick for retrieving the value from a key as for retrieving the keys from a value without using up twice as much memory, since they share the same place in my hash table. I only need to add a reference to another Entry.

That said, I will use your code for benchmarking. =)

**Comment by bjornregnell** [comment]
Aha, yes your java implementation should takes less memory. Have you tested on some real data how much space the refs take? I guess in my Scala solution it's "only" refs that are the overhead, as the actual data values are not duplicated... if I'm right? Would be interesting to see actual tests on some data sets...

**Comment by avatar1337** [comment]
Yes, only refs. The data is never duplicated, so my solution should still take up less memory. When getKeys(value) is called, the method constructs a List (because it was faster than a Set, they must be unique anyway) and returns it.

```
	public Collection<K> getKeys(Object value) {		
		return findKeys(index((V)value), (V)value);
	}

	private int index(Object o) {
		int index = o.hashCode() % table.length;
		if (index < 0 )	index += table.length;
		return index;
	}

	private Collection<K> findKeys(int index, V value) {
		return findKeys(table, index, value);
	}
	
	private Collection<K> findKeys(Entry<K,V>[] table, int index, V value) {
		Collection<K> keys = new LinkedList<>();
		
		Entry<K,V> e = table[index];
		while(e != null && e.shadow != null) {
			if(e.shadow.value.equals(value)) keys.add(e.shadow.key);
			e = e.next;
		}
		return keys;
	}
```
Will your `private val shadow =  scala.collection.mutable.Map.empty[V,Set[K]]` only contain references to already created entries? I'm not sure how it looks underhood. If scala uses a HashMap I believe it creates Entry objects and put them in a table. Even if you put references in your map, it should still create Entry objects which will allocate more memory, right? My ShadowHashMap uses the same Entry objects underneath and only stores references to them.

EDIT: Nevermind, I saw you answered that in your edited version above.

Do you know of a way to measure how much memory my map uses on JVM? In C++ I would use sizeof.

**Comment by bjornregnell** [comment]
> Will your private val shadow = scala.collection.mutable.Map.empty[V,Set[K]] only contain references to already created entries?

Yes, that's how I understand it; Set is immutable and the refs can thus be shared.

Also, all your Entry instances actually have to reserve memory for twice as many pointers even if they are never used, and most of them might be null, but each instance still requires some bytes each as a placeholder for those shadow and next refs... 

The Scala solution in *my second* comment (counting from top) with a memoized reverse mapping only allocates extra ref space when really asked for, and also reuses the referenced instances, if I'm right. So this really needs some benchmarking to investigate the trade-off in terms of ratio of value duplicates over a specific key set.

> Do you know of a way to measure how much memory my map uses on JVM? In C++ I would use sizeof.

https://stackoverflow.com/questions/9368764/calculate-size-of-object-in-java

https://stackoverflow.com/questions/52353/in-java-what-is-the-best-way-to-determine-the-size-of-an-object?noredirect=1&lq=1

**Comment by avatar1337** [comment]
I will do more testing and more accurately later, but I tested your ReversableMapBuilder against my ShadowHashMap. I added 1 million integer keys (in order) and random integers between 0 and 999999 as values and my ShadowHashMap seems to be consistently ~1.7 times faster in retrieving all the keys. =)

// N = 1 000 000
for(i <- 0 until N) rmb.keysOf(i) vs. for(int i=0; i < N; i++) numbers.getKeys(i);

133 ns vs 226 ns in mean time.

This could change however after proper warm up.

**Comment by bjornregnell** [comment]
Interesting! Would be nice to see what hapens if you only need to retrieve 50% of the keys. I gues in many use cases not all key-values will needed in a reverse lookup. Would alse be interseting to see how the memoized, immutable case class performs...

**Comment by avatar1337** [comment]
For only 50% retrieval mine is ~1.4-1.5 times faster, about the same for 20%.
The fewer unique values I have in my ShadowHashMap the more key retrieval should go towards O(n) though. Value retrieval time isn't affected.

```
 ✤ HASHTABLE CONTENT ✤
⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺⎺
0: 〔0 = 2〕
1: 〔1 = 2, SHADOW ➜ 18 = 1 @ index: 18〕
2: 〔2 = 3, SHADOW ➜ 0 = 2 @ index: 0〕 ➜ 〔SHADOW ➜ 1 = 2 @ index: 1〕 ➜ 〔SHADOW ➜ 3 = 2 @ index: 3〕 ➜ 〔SHADOW ➜ 4 = 2 @ index: 4〕 ➜ 〔SHADOW ➜ 6 = 2 @ index: 6〕 ➜ 
〔SHADOW ➜ 8 = 2 @ index: 8〕
3: 〔3 = 2, SHADOW ➜ 2 = 3 @ index: 2〕 ➜ 〔SHADOW ➜ 7 = 3 @ index: 7〕 ➜ 〔SHADOW ➜ 15 = 3 @ index: 15〕
4: 〔4 = 2, SHADOW ➜ 10 = 4 @ index: 10〕 ➜ 〔SHADOW ➜ 12 = 4 @ index: 12〕 ➜ 〔SHADOW ➜ 13 = 4 @ index: 13〕 ➜ 〔SHADOW ➜ 17 = 4 @ index: 17〕
5: 〔5 = 5, SHADOW ➜ 5 = 5 @ index: 5〕 ➜ 〔SHADOW ➜ 9 = 5 @ index: 9〕 ➜ 〔SHADOW ➜ 11 = 5 @ index: 11〕 ➜ 〔SHADOW ➜ 14 = 5 @ index: 14〕 ➜ 〔SHADOW ➜ 16 = 5 @ index: 
16〕 ➜ 〔SHADOW ➜ 19 = 5 @ index: 19〕
6: 〔6 = 2〕
7: 〔7 = 3〕
8: 〔8 = 2〕
9: 〔9 = 5〕
10: 〔10 = 4〕
11: 〔11 = 5〕
12: 〔12 = 4〕
13: 〔13 = 4〕
14: 〔14 = 5〕
15: 〔15 = 3〕
16: 〔16 = 5〕
17: 〔17 = 4〕
18: 〔18 = 1〕
19: 〔19 = 5〕
20: 〔〕
21: 〔〕
22: 〔〕
23: 〔〕
24: 〔〕
25: 〔〕
26: 〔〕
27: 〔〕
28: 〔〕
29: 〔〕
30: 〔〕
31: 〔〕

```
For instance, here is my hash table with 20 integer keys but only 5 values (1-5). You get all shadow pointers between 1-5. So you get **v** number of linked lists where **v** is the number of unique values. This seems to be a fatal blow to my ShadowHashMap. It becomes much slower. My map is only better if there is a lot of unique values.

I tried your memoized version and it is incredibly slow hehe. I had to go down to only 10000 keys instead of a million, and even then the mean retrieval time is 270363 ns.
